### PR TITLE
Video: add spacing block supports

### PR DESCRIPTION
**Pull request by ramonjd**
Related:
- https://github.com/WordPress/gutenberg/issues/43241
- https://github.com/WordPress/gutenberg/issues/43243

## What?

Enabling spacing support for the Video block

## Why?

To create consistency across blocks.

## How?

Adding the relevant block supports in block.json

## Testing Instructions

1. Load the block editor and add a Video block. Add some content or use the example block code below.
2. Confirm that the dimensions control panel is there
3. Add margin and padding.
4. Save and confirm the styles appear on the frontend.
5. Test the new support works for the block via theme.json and global styles. See example JSON below.

```json
	"styles": {
		"blocks": {
			"core/video": {
				"spacing": {
					"margin": {
						"top": "30px",
						"bottom": "40px"
					}
				}
			}
		}
	}
```

https://user-images.githubusercontent.com/6458278/185363765-a2a8bdf5-2318-4b36-89a5-585359a739cf.mp4

### CODE REVIEW

**Review by apeatling — CHANGES_REQUESTED** [review]
Same issue as the spacer block here, everything works except for the margin settings in theme.json. It seems to be an issue with splitting out margin sides. If I set margin and padding as a single value then it works.

**Comment by ramonjd** [comment]
@apeatling Thanks for testing. 

The styles are printed out in the CSS, but it looks like the layout styles `margin-block-start` and `margin-block-end` are taking precedence

Screen Shot 2022-08-19 at 8 41 40 am

Not sure how to tackle it yet, maybe there has to be an exception in layout.php if it detects these margin side values.

cc @andrewserong for the usual sanity check 😄

**Comment by ramonjd** [comment]
Looks like it's coming from the [theme.json layout rules](https://github.com/WordPress/gutenberg/blob/04d57b34b8574e1f25de0f0bc5862e3da045c6ac/lib/compat/wordpress-6.1/theme.json#L229) printed out here:

https://github.com/WordPress/gutenberg/blob/c76c87e9f1ec22578436d8e24969a29c568c9cd0/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php#L1369-L1368

Naive way forward might be to check for existing values and compare with $declarations rules, but given that there could be any rule in there, it's far from bulletproof 
🤔

**Comment by andrewserong** [comment]
Oh, good discussion here!

Margin is a tricky one — from the testing I did on a couple of PRs yesterday, I think it's pretty risky to attempt to reduce the specificity of the spacing between blocks in order for the `theme.json` block-level margin to be able to take precedence. I believe the important thing is that the margin set at the individual block level in post content works as expected.

From my perspective, I _think_ the spacing of the wrapper that a block is in should take precedence over block-level-in-theme.json margin, and block level in individual block in post content should take precedence over the wrapper block's spacing. Otherwise we risk breaking how block gap works, as there's an inherent friction between the concepts of these two features.

One of the workarounds that folks can use if they want the `theme.json` value to work, is to wrap their block in another wrapper so that they avoid the spacing rules from being applied.

I suppose if I were to pick between the two, I'd pick _margin_ to behave slightly unpredictably (as that is kinda how margin tends to work?) rather than trying to get the block spacing feature to work around it. Another thing to keep in mind is that left/right margin is already overwritten by alignment styles anyway, so there's precedent for margin only working in certain circumstances.

What do you think? Is there another avenue that either of you think we should pursue for how the layout spacing works here?

**Comment by ramonjd** [comment]
> I suppose if I were to pick between the two, I'd pick margin to behave slightly unpredictably (as that is kinda how margin tends to work?) rather than trying to get the block spacing feature to work around it. 

Thanks for the input @andrewserong 

After having a look at the code I think I agree 😄 It's a bit of a tangled web. 

Do you think we could get this and the [spacer PR](https://github.com/WordPress/gutenberg/pull/43366#pullrequestreview-1077774759) in given that block supports in the editor work?

> One of the workarounds that folks can use if they want the theme.json value to work, is to wrap their block in another wrapper so that they avoid the spacing rules from being applied.

I tried this before, but couldn't get around the layout styles:

Screen Shot 2022-08-19 at 10 17 51 am

**Comment by andrewserong** [comment]
> Do you think we could get this and the https://github.com/WordPress/gutenberg/pull/43366#pullrequestreview-1077774759 in given that block supports in the editor work?

Yes, it's great that we've found this edge case, but I don't think it should block these opt-in PRs, personally.

> I tried this before, but couldn't get around the layout styles:

Ah, good catch — I think in another PR I must have been testing with a block that was in a Row or Stack, possibly, that doesn't have those spacing rules. I reckon it might be worth us opening an issue to see if it's worth coming up with a solution for the Group block — possibly an "unset" setting for spacing / block gap, or a layout type that is "no layout"? At any rate, it'd be good to look into separately to verify whether it's something that in practice folks will use, or if we're mostly encountering it here because we're testing these features in detail 🤔

Possibly a more likely real world use case for block-level margin in `theme.json` is themes that opt out of block styles altogether with `settings.spacing.blockGap` set to `null`?

File: packages/block-library/src/video/block.json [hunk 1]
```diff
@@ -77,6 +77,10 @@
 	"supports": {
 		"anchor": true,
 		"align": true,
+		"spacing": {
+			"margin": true,
+			"padding": true
```
Comment (andrewserong) [inline packages/block-library/src/video/block.json:82]: I noticed a slight issue with padding in TwentyTwentyTwo that I think is caused by the Video block using a `width: 100%` rule, in that the padding winds up being outside the `100%` as opposed to part of it. Here's a Video block with padding next to a Group block with padding:

image

Shall we add a `box-sizing: border-box` rule as in the Audio block PR (It looks like most blocks with padding have the rule added, I think, along with a comment explaining why the rule was required) https://github.com/WordPress/gutenberg/pull/43351#pullrequestreview-1076698428

Comment (ramonjd) [reply packages/block-library/src/video/block.json:82]: Thanks for finding that. **TIL**

I'll add the rule and retest. 🙇

Comment (ramonjd) [reply packages/block-library/src/video/block.json:82]: Looks like that did the trick. Thank you!!!!
Screen Shot 2022-08-19 at 12 30 28 pm

**Comment by ramonjd** [comment]
> I reckon it might be worth us opening an issue to see if it's worth coming up with a solution for the Group block — possibly an "unset" setting for spacing / block gap, or a layout type that is "no layout"? At any rate, it'd be good to look into separately to verify whether it's something that in practice folks will use, or if we're mostly encountering it here because we're testing these features in detail

Thanks @andrewserong I'll add a TODO note for myself to make an issue today or next week.

**Review by andrewserong — APPROVED** [review]
LGTM! Thanks for the update. We might need Andy to re-approve, too, before Github will let us merge 🙂

**Review by apeatling — APPROVED** [review]
Thanks for the follow ups!

**Pull request merged**